File: java/src/realtimeweb/redditservice/domain/Post.java
package realtimeweb.redditservice.domain;

import java.math.BigDecimal;
import java.util.Map;




/**
 * A link (or self-text) that has been submitted to Reddit.
 */
public class Post {
	
    // prepend reddit.com to this
    private String permalink;
    private String author;
    private String title;
    private Integer downs;
    // epoch time, save as long, convert to date
    private Long created;
    private String subreddit;
    // if is self then use data->url
    private String content;
    private Boolean isSelf;
    private String id;
    private Integer ups;
    private Boolean isNsfw;
    
    
    /*
     * @return A permanent url that directs to this Post.
     */
    public String getPermalink() {
        return this.permalink;
    }
    
    /*
     * @param A permanent url that directs to this Post.
     * @return String
     */
    public void setPermalink(String permalink) {
        this.permalink = permalink;
    }
    
    /*
     * @return The username of the author of this Post.
     */
    public String getAuthor() {
        return this.author;
    }
    
    /*
     * @param The username of the author of this Post.
     * @return String
     */
    public void setAuthor(String author) {
        this.author = author;
    }
    
    /*
     * @return The title of this Post.
     */
    public String getTitle() {
        return this.title;
    }
    
    /*
     * @param The title of this Post.
     * @return String
     */
    public void setTitle(String title) {
        this.title = title;
    }
    
    /*
     * @return The number of downvotes associated with this Post.
     */
    public Integer getDowns() {
        return this.downs;
    }
    
    /*
     * @param The number of downvotes associated with this Post.
     * @return Integer
     */
    public void setDowns(Integer downs) {
        this.downs = downs;
    }
    
    /*
     * @return The date that this Post was created.
     */
    public Long getCreated() {
        return this.created;
    }
    
    /*
     * @param The date that this Post was created.
     * @return Integer
     */
    public void setCreated(Long created) {
        this.created = created;
    }
    
    /*
     * @return The subreddit that this Post was made in.
     */
    public String getSubreddit() {
        return this.subreddit;
    }
    
    /*
     * @param The subreddit that this Post was made in.
     * @return String
     */
    public void setSubreddit(String subreddit) {
        this.subreddit = subreddit;
    }
    
    /*
     * @return The text of the post, or a url if it is not a self Post.
     */
    public String getContent() {
        return this.content;
    }
    
    /*
     * @param The text of the post, or a url if it is not a self Post.
     * @return String
     */
    public void setContent(String content) {
        this.content = content;
    }
    
    /*
     * @return Whether or not this Post was text (True), or a URL (False).
     */
    public Boolean getIsSelf() {
        return this.isSelf;
    }
    
    /*
     * @param Whether or not this Post was text (True), or a URL (False).
     * @return Boolean
     */
    public void setIsSelf(Boolean isSelf) {
        this.isSelf = isSelf;
    }
    
    /*
     * @return A unique ID for this Post. A combination of letters, numbers, and dashes.
     */
    public String getId() {
        return this.id;
    }
    
    /*
     * @param A unique ID for this Post. A combination of letters, numbers, and dashes.
     * @return String
     */
    public void setId(String id) {
        this.id = id;
    }
    
    /*
     * @return The number of upvotes associated with this Post.
     */
    public Integer getUps() {
        return this.ups;
    }
    
    /*
     * @param The number of upvotes associated with this Post.
     * @return Integer
     */
    public void setUps(Integer ups) {
        this.ups = ups;
    }
    
    /*
     * @return Whether or not this Post is Not Safe for Work (NSFW).
     */
    public Boolean getIsNsfw() {
        return this.isNsfw;
    }
    
    /*
     * @param Whether or not this Post is Not Safe for Work (NSFW).
     * @return Boolean
     */
    public void setIsNsfw(Boolean isNsfw) {
        this.isNsfw = isNsfw;
    }
    
	
	/**
	 * Creates a string based representation of this Post.
	
	 * @return String
	 */
	public String toString() {
		return "Post[" +permalink+", "+author+", "+title+", "+downs+", "+created+", "+subreddit+", "+content+", "+isSelf+", "+id+", "+ups+", "+isNsfw+"]";
	}
	
	/**
	 * Internal constructor to create a Post from a json representation.
	 * @param map The raw json data that will be parsed.
	 * @return 
	 */
    public Post(Map<String, Object> raw) {
        // TODO: Check that the data has the correct schema.
        // NOTE: It's much safer to check the Map for fields than to catch a runtime exception.
        try {
            this.permalink = ((Map<String, Object>) raw.get("data")).get("permalink").toString();
            this.author = ((Map<String, Object>) raw.get("data")).get("author").toString();
            this.title = ((Map<String, Object>) raw.get("data")).get("title").toString();
            this.downs = Integer.parseInt(((Map<String, Object>) raw.get("data")).get("downs").toString());
            this.created = new BigDecimal(((Map<String, Object>) raw.get("data")).get("created").toString()).longValue();
            this.subreddit = ((Map<String, Object>) raw.get("data")).get("subreddit").toString();
            this.content = ((Map<String, Object>) raw.get("data")).get("selftext").toString();
            this.isSelf = Boolean.parseBoolean(((Map<String, Object>) raw.get("data")).get("is_self").toString());
            this.id = ((Map<String, Object>) raw.get("data")).get("id").toString();
            this.ups = Integer.parseInt(((Map<String, Object>) raw.get("data")).get("ups").toString());
            this.isNsfw = Boolean.parseBoolean(((Map<String, Object>) raw.get("data")).get("over_18").toString());
        } catch (NullPointerException e) {
    		System.err.println("Could not convert the response to a Post; a field was missing.");
    		e.printStackTrace();
    	} catch (ClassCastException e) {
    		System.err.println("Could not convert the response to a Post; a field had the wrong structure.");
    		e.printStackTrace();
        }
    
	}	
}

File: python/redditservice/post.py
class Post(object):
	"""
	A link (or self-text) that has been submitted to Reddit.
	"""
	def __init__(self, ups, downs, created, subreddit, id, title, author, is_self, is_nsfw, content, permalink):
		"""
		Creates a new Post
        :param self: This object
        :type self: Post
        :param ups: The number of upvotes associated with this Post.
        :type ups: int
        :param downs: The number of downvotes associated with this Post.
        :type downs: int
        :param created: The date that this Post was created.
        :type created: int
        :param subreddit: The subreddit that this Post was made in.
        :type subreddit: string
        :param id: A unique ID for this Post. A combination of letters, numbers, and dashes.
        :type id: string
        :param title: The title of this Post.
        :type title: string
        :param author: The username of the author of this Post.
        :type author: string
        :param is_self: Whether or not this Post was text (True), or a URL (False).
        :type is_self: boolean
        :param is_nsfw: Whether or not this Post is Not Safe for Work (NSFW).
        :type is_nsfw: boolean
        :param content: The text of the post, or a url if it is not a self Post.
        :type content: string
        :param permalink: A permanent url that directs to this Post.
        :type permalink: string
        :returns: Post
		"""
		self.ups = ups
		self.downs = downs
		self.created = created
		self.subreddit = subreddit
		self.id = id
		self.title = title
		self.author = author
		self.is_self = is_self
		self.is_nsfw = is_nsfw
		self.content = content
		self.permalink = permalink
	
	@staticmethod
	def _from_json(json_data):
		"""
		Creates a Post from json data.
        :param json_data: The raw json data to parse
        :type json_data: dict
        :returns: Post
		"""
		return Post(json_data['data']['ups'],
					json_data['data']['downs'],
					json_data['data']['created'],
					json_data['data']['subreddit'],
					json_data['data']['id'],
					json_data['data']['title'],
					json_data['data']['author'],
					json_data['data']['is_self'],
					json_data['data']['over_18'],
					json_data['data']['selftext'] if json_data['data']['is_self'] else json_data['data']['url'],
					json_data['data']['permalink'])